src/txn/txn_timestamp.c

```diff
@@ -1047,7 +1105,15 @@ __wt_txn_publish_commit_timestamp(WT_SESSION_IMPL *session)
 	txn = &session->txn;
 	txn_global = &S2C(session)->txn_global;
 
-	if (F_ISSET(txn, WT_TXN_PUBLIC_TS_COMMIT))
+	/*
+	 * If we know for a fact that this is a prepared transaction, don't
+	 * add to the durable queue. If we poll all_durable after setting the
+	 * commit timestamp of a prepared transaction, that prepared transaction
+	 * should NOT be visible.
+	 */
+	if (F_ISSET(txn, WT_TXN_PREPARE))
+		return;
```

Comment:
Do we need to remove an entry from the durable queue when preparing a transaction?

Reply:
The sequence for a prepared transaction must be:
- Set prepare ts.
- Set commit ts.
- Set durable ts.
- Set more commit ts (optional).

This is to protect us from inserting into the queue at either commit timestamp. The first one will hit this condition. Any that come after the durable timestamp will hit the `WT_TXN_TS_PUBLISHED` check further up.